3880 z managed integration improvements

adrianfilip opened the pull request:
Hi @jdegoes,

This is for #3880.
Let me know if I anything should be changed!

iravid commented:
Hi @adrianfilip, this looks pretty cool and a good addition. I have a few suggestions:
1. The case classes can be placed outside of the platform-specific trait. Only the constructors need to be there.
2. It'd be nice if these were not case classes but rather traits or abstract classes, such that we can later implement them using a backing `ZStream` in `ZStream#toInputStream` or a future `ZSink#toOutputStream`.

iravid commented on core/jvm/src/main/scala/zio/ZManagedPlatformSpecific.scala:
```diff
@@ -0,0 +1,132 @@
+            case 0 => None
+            case _ => {
+              val buffer = new java.io.ByteArrayOutputStream();
+              val data   = new Array[Byte](4096);
```
The chunk size here should be configurable

iravid commented on core/jvm/src/main/scala/zio/ZManagedPlatformSpecific.scala:
```diff
@@ -0,0 +1,132 @@
+ * A functional wrapper over a java.io.InputStream.
+ */
+private[zio] case class InputStream private (private val is: java.io.InputStream) extends ZInputStream {
+  def readN(n: Int): ZIO[Blocking, IOException, Option[Chunk[Byte]]] =
```
With this signature, there's no way for the caller to infer if the input stream is exhausted. How about encoding this as `ZIO[Blocking, Option[IOException], Chunk[Byte]]`, where `Some(IOException)` is an actual error, `None` is an end-of-stream signal and a value is a chunk?

I would also suggest to make the non-blocking read optional (using a flag). If the flag is false, don't check `available` before reading and just let the underlying input stream manage that.

adrianfilip replied:
Great suggestion. I think we should remove the available check alltogether if we use the encoding you proposed. And have each implementation decide how it wants to deal with readN on an exhausted stream. I think it makes things simpler.

 What do you think?

iravid commented on core/jvm/src/main/scala/zio/ZManagedPlatformSpecific.scala:
```diff
@@ -0,0 +1,132 @@
+  def skip(n: Long): ZIO[Blocking, IOException, Long] =
+    effectBlocking(is.skip(n)).refineToOrDie[IOException]
+
+  def readAll: ZIO[Blocking, IOException, Option[Chunk[Byte]]] =
```
This can be `ZIO[Blocking, IOException, Chunk[Byte]]` and the blocking can be configured through a flag.

adrianfilip replied:
I understand. Looking at ZIO[Blocking, Option[IOException], Chunk[Byte]] from readN I think this place would be a good one to use the same encoding and also drop the available check alltogether (we already get the info from the read return code).

iravid commented on core/jvm/src/main/scala/zio/ZManagedPlatformSpecific.scala:
```diff
@@ -0,0 +1,132 @@
+  def write(chunk: Chunk[Byte]): ZIO[Blocking, IOException, Unit] =
+    effectBlocking {
+      os.write(chunk.toArray)
+      os.flush()
```
Should we always flush?

adrianfilip replied:
Good point. No, we don't.  I was thinking about treating write + flush as "atomic" in a write, but it doesn't have to be and it would actually reduce performance (because of all that IO). I will remove the flush.

adrianfilip commented:
@iravid Sounds great. I'll make the changes.

adrianfilip commented on core-tests/jvm/src/test/scala/zio/ZManagedPlatformSpecificSpec.scala:
```diff
@@ -0,0 +1,163 @@
+
+object ZManagedPlatformSpecificSpec extends ZIOBaseSpec {
+
+  val target = new ZManagedPlatformSpecific {}
```
I can't access it like you mention.

iravid replied:
Oh right. We need to modify the ZManaged companion object to extend the ZManagedPlatformSpecific trait.

adrianfilip commented:
@iravid I implemented the changes based on your suggestions. I updated my approach as a reply to your comments.
Let me know if this is ok and also if the new ZInputStream and ZOutputStream files should be moved elsewhere.

(The failing tests are unrelated to the PR (i raised a ticket for that https://github.com/zio/zio/issues/3908).)

adrianfilip commented:
@jdegoes @iravid I think it's good to go now. Let me know if you think I should change anything.
